community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementationHandle.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api;
+
+import org.neo4j.kernel.api.KernelTransactionHandle;
+import org.neo4j.kernel.api.exceptions.Status;
+
+/**
+ * A {@link KernelTransactionHandle} that wraps the given {@link KernelTransactionImplementation}.
+ * This handle knows that {@link KernelTransactionImplementation}s can be reused and represents a single logical
+ * transaction. This means that methods like {@link #markForTermination(Status)} can only terminate running
+ * transaction this handle was created for.
+ */
+class KernelTransactionImplementationHandle implements KernelTransactionHandle
+{
+    private final long txReuseCount;
+    private final KernelTransactionImplementation tx;
+
+    KernelTransactionImplementationHandle( KernelTransactionImplementation tx )
+    {
+        this.txReuseCount = tx.getReuseCount();
+        this.tx = tx;
+    }
+
+    @Override
+    public long lastTransactionTimestampWhenStarted()
+    {
+        return tx.lastTransactionTimestampWhenStarted();
```

Comment:
This will leak timestamp from potentially not opened/ reused transaction. It breaks single logical transaction principle.